rpc/client.go

```diff
@@ -91,219 +97,224 @@ type Client struct {
-}
+	go func() {
+		newCall.Error = c.conn.Load().(internalConn).client.Call(serviceMethod, args, reply)
 
-// IsConnected returns whether the client is connected.
-func (c *Client) IsConnected() bool {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.Client != nil
-}
+		if newCall.Error != nil {
+			// From net/rpc/client.go; remote hung up. Our heartbeat runloop will
+			// reconnect, so this is retryable.
+			if newCall.Error == io.ErrUnexpectedEOF {
+				newCall.Error = newRPCError(newCall.Error)
+			}
+		}
+		newCall.Done <- newCall
+	}()
 
-// IsHealthy returns whether the client is healthy.
-func (c *Client) IsHealthy() bool {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.healthy
+	return newCall
 }
 
-// Addr returns remote address of the client.
-func (c *Client) Addr() net.Addr {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.addr
-}
+func (c *Client) connect() error {
+	conn, err := tlsDialHTTP(c.addr.NetworkField, c.addr.StringField, c.tlsConfig)
+	if err != nil {
+		return err
+	}
+	c.conn.Store(internalConn{
+		conn:   conn,
+		client: rpc.NewClientWithCodec(codec.NewClientCodec(conn)),
+	})
 
-// LocalAddr returns the local address of the client.
-func (c *Client) LocalAddr() net.Addr {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.lAddr
+	return nil
 }
 
-// RemoteOffset returns the most recently measured offset of the client clock
-// from the remote server clock.
-func (c *Client) RemoteOffset() proto.RemoteOffset {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.offset
+// Healthy returns a channel that is closed when the client becomes healthy.
+func (c *Client) Healthy() <-chan struct{} {
+	return c.healthy.Load().(chan struct{})
 }
 
-// Close removes the client from the clients map and closes
-// the Closed channel.
+// Close closes the Closed channel, which triggers the end of the run loop and
+// removal from the clients map.
 func (c *Client) Close() {
 	clientMu.Lock()
-	if !c.closed {
-		if c.cached {
-			delete(clients, c.Addr().String())
-		}
-		c.mu.Lock()
-		c.healthy = false
-		c.closed = true
-		if c.Client != nil {
-			c.Client.Close()
-		}
-		c.mu.Unlock()
-		close(c.Closed)
+	defer clientMu.Unlock()
+
+	if c.isClosed {
+		return
 	}
-	clientMu.Unlock()
+
+	c.isClosed = true
+	delete(clients, c.addr)
+	close(c.Closed)
 }
 
 // runHeartbeat sends periodic heartbeats to client. Closes the
 // connection on error. Heartbeats are sent in an infinite loop until
 // an error is encountered.
-func (c *Client) runHeartbeat(stopper *stop.Stopper) {
-	if log.V(2) {
-		log.Infof("client %s starting heartbeat", c.Addr())
+func (c *Client) runHeartbeat(retryOpts retry.Options) {
+	isHealthy := false
+	setHealthy := func() {
+		if isHealthy {
+			return
+		}
+		isHealthy = true
+		close(c.healthy.Load().(chan struct{}))
+	}
+	setUnhealthy := func() {
+		if isHealthy {
+			isHealthy = false
+			c.healthy.Store(make(chan struct{}))
+		}
 	}
 
+	// initial connection
+	connErr := c.connect()
```

Comment:
If the initial connection fails, we don't log its error and we retry immediately without waiting for the initial backoff. What if instead we started with `connErr := errors.New("not started yet")`? I think that would do the right thing (i.e. an immediate connection attempt inside the loop, and every attempted connection failure would be logged and retried on schedule).